README.md
# Wind Noise Detection in Audio Files
### University Of Salford Acoustics Research Centre

# WindNoiseDetection

A fork of [kenders2000/WindNoiseDetection](https://github.com/kenders2000/WindNoiseDetection). I have just modified the wrappper to allow for reading file locations and names from a file and implemented MPI to allow for parallel processing.

*Warning* there lurks a memory leak somewhere, so running a large numbers of files will eventually lead to the program crashing and you'll have to start again. 

To run you'll need cygwin or a bash terminal with MPI installed. 

To compile run 

```
make winDDetFile
```

or if you don't want to use the list of files feature just `make windDet`

To run:
```
mpirun -np 19 windDetFile.exe -i paths/ -o /cygdrive/d/WindDetection/NL_outputs/errors/

```

where -i is the argument for the location of your file paths and -o is the output directory and -np is the number of processors to run with. 

See below for original README and details.



## Overview

This program automatically analyses .wav files and detects regions where there may be microphone wind noise.  Wind noise is known to degrade the quality of outdoor recordings and causes problems with hearing aids.  

This tool could be used when there are a large number of recordings of a particular source, for example clips of recordings from mobile devices of an outdoor concert (user generated content).  This tool can quickly analyse the audio from each to determine samples with the least amount of wind noise and therefore of the highest quality.

The wind noise detection algorithm also provides time stamps identifying the regions which are free of wind noise.  This can allow producers to quickly stitch together content from multiple sources, or flag regions where extra processing such as noise reduction is required.

Perceptual tests have shown that when wind noise is present, the Signal to Noise ratio between the wanted signal and the wind noise is a good indicator of the quality.  Therefore this detector works by first trying to determine the presence of wind noise and then estimating the SNR to provide an indication of the possible degradation to the audio quality.

_See Iain R. Jackson, Paul Kendrick, Trevor J. Cox, Bruno M. Fazenda and Francis Li, "Perceptual evaluation of the functional and aesthetic degradation of speech by wind noise during recording", Proceedings of Meetings on Acoustics, ASA, Vol. 19, 060170 (2013) for further information._


## Usage

The program requires compiling using a C++ compiler and runs as a command line executable which analyses a wavfile and provides information as regards the level, location and effect on quality of any wind noise that may be present.

Wav files must be 16 or 32 bit int, or 32 bit float PCM. All channels are collapsed to one (average of all channels)

```
windDet -i wav_filename -o output_filename -j  output_filename_json[-t 1] [-g 1] [-f 43] [-w 25] [-v 0]
     -i and -o are required parameters, they provide the input .wav filename and the output filename respectively. Specify the output as a text file, include path if required, the program will also output a json file with the same name in the same path.
	 -j specifies the path and name of the json file containing the output data.
     -t n, sets the tree size to n, where n is an integer (1,2 or 3). The default is 2, 1 is fastest but is least accurate.
     -g 1, sets the gain to apply. The default is 1.  If you have calibrated audio you can adjust the level, by default the peak pressure level is assumed to 104 dB, gain can be applied if you have an accurate calibration of the pressure level on the recording, which may increase accuracy but is not expressly required.
     -f n, sets the number of frames used to produce the analysis window to n, windows are 1024 samples long.  The default is n = 43 frames, which is about 1 s
	 -w sets the threshold degradation in quality level for identifying noise free segments, default is 25 (%), but this can be adjusted depending on the required quality level for the application
	 -v tells the program weather or not to print out the terminal output, verbose is 1 or 0, 1 for printout info 0 for no printout.
   
```
   
### Output file description and example

The output text file will contain three sets of outputs: global statistics, wind noise time history, and start and end points of wind noise. Using the following parameter set should provide the results presented.  Additionally a json file will be oututted with the same data


```
windDet.exe -i test/iphone1.wav -o test/iphone1.txt -j test/iphone1.json -t 1 -g 1 -f 43 -w 25 -v 0
```

#### Global statistics about  wind noise

This shows the percentage  of frames which are either free of wind noise or contain wind noise within a particular Signal to (wind)Noise ratio (SNR) range.  These global statistics can be used to make a decision as regards the usability of the file.  There are three rows of numbers. 

 1. 5 ranges of SNRs. 
 2. Corresponding degradation to quality for a particular SNR range.
 3. Percentage of time the signal is within a particular Quality range.
 
The test output of the text file shows an example where 36.4% of frames are free of wind noise, but 54% of the time the quality is degraded by over 50%.  (the json outputs the same data but formatted as to be readable by a json interpreter).


| SNR | >20dB | 20 to 10dB   | 10 to 0dB   | 0 to -10dB | -10 to -20dB   | <-20dB |
|--- | ----- | ---        | ---        | ---        | ---          | ---    |
| **Degradation** | <24%  | 24 to 40   | 40 to 54%   | 54 to 69%    | 69 to 95%  | >96%   |
| **Total time** | 36.4%  | 0.0%       | 18.2%       | 27.3%        | 9.1%         |  9.1%   |

These statistics can be used to make quick judgements about the file regarding the distribution of wind noise levels, but it is down to the required application to choose a suitable set of criteria for acceptability.

#### Wind noise time history

The next set of parameters is the wind noise time history.  Three columns are presented: 

 1. The time at the end of the analysis window
 2. The predicted quality degradation due to wind noise, which is computed from  the predicted SNR
 3. An A-weighted Leq over the window.  
    
The algorithm assumes that a maximum sample value is equivalent to a Sound Pressure Level of 104 dB, [this value was taken as an average of the peak undistorted level of a number of common recording devices](http://usir.salford.ac.uk/29371/1/Distortion_of_Portable_Audio_Devices.pdf). This level can be adjusted by varying the input gain using `-g` should the estimated dBA level appear unrealistic.


| T(s)  | Quality Degradation(%)  | dBA |
| ----- | -------------------------- | --- |
| 1.0 | 0           |  65 |
| 2.0 | 0           |  79 |
| 3.0 | 34            |  82 |
| 4.0 | 0           |  74 |
| 5.0 | 0           |  79 |
| 6.0 | 50            |  88 |
| 7.0 | 66            |  87 |
| 8.0 | 45            |  85 |
| 9.0 | 52            |  87 |
|10.0 | 83            |  87 |
|11.0 | 0           |  82 |

#### Start and end points of wind noise

Finally a set of times is returned which represent the start and end points, in seconds, of regions in the recording which are free of wind noise.  So the following example shows two segments, one from 0 to 1.99s and the next from 2.99 to 4.99 s.

```
0.00  1.99
2.99   4.99
```

An intermediate file is also created which contains the raw classes for sanity checking.

## The Good Recording Project - overview

The University of Salford Acoustics Research Centre is carrying out a project entitled [The Good Recording Project](http://www.goodrecording.net/), investigating how recording errors can affect the perceived quality of audio.  In particular with regards to content recorded on mobile device such as mobile phones, video cameras and other such media often known as "user generated content".  This wind noise detector was developed as part of an [EPSRC funded project](http://gow.epsrc.ac.uk/NGBOViewGrant.aspx?GrantRef=EP/J013013/1).

It is licensed under the MIT license and uses the [KISS_fft library](http://sourceforge.net/projects.kissfft/) which is licensed under the revised BSD license. You are therefore free to use this in pretty much any project.  If you do use this work in any project [an acknowledgement would be appreciated](http://www.salford.ac.uk/computing-science-engineering/subjects/acoustics-audio-and-video).

The software was developed at the University of Salford and uses a bagged decision tree to analyse MFCCs to discover if an audio frame contains microphone wind noise, it is part of the Good Recording Project at the University of Salford Acoustic Research Centre. 


## How does it work?

First the 1024 samples of the wav file are read in, the level calibrated roughly, assuming that a peak sample value is equivalent to 104 dB.  Then 16 Mel Frequency Cepstrum Coefficients (20 to 8000 Hz) are extracted (Hanning weighting), (MFCCs) this is a spectral representation of the frame.  Three sets of frames are collected and first 12 coefficients stored; the first coefficient is replaced with the A-weighted level over the frame in dB.  These 36 features are concatenated and then two ensembles of decision trees take these features as an input and output, a class number.  One tree-ensemble attempts to estimate the wind level:

The Average Wind Level class varies from 0 to 3.

```

history.txt
dir
ls
cd /cygdrive/d/TMP_R_Clones/WindNoiseDetection/
ls
make
make windDet
g++
man g++ 
g++ --version
make clean
make windDet
windDet.exe -i test/iphone1.wav -o test/iphone1.txt -j test/iphone1.json -t 1 -g 1 -f 43 -w 25 -v 0
./windDet.exe -i test/iphone1.wav -o test/iphone1.txt -j test/iphone1.json -t 1 -g 1 -f 43 -w 25 -v 0
 ./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211005_NapkenLk_duskdawn/20211005T023500-0400_SS.wav' -o test/sampletest.txt -j test/sampletest.json -t 1 -g 1 -f 43 -w 25 -v 1
make windDet
 ./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211005_NapkenLk_duskdawn/20211005T023500-0400_SS.wav' -o test/sampletest.txt -j test/sampletest.json -t 1 -g 1 -f 43 -w 25 -v 0
 ./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211005_NapkenLk_duskdawn/20211005T124200-0400_SR.wav' -o test/sampletest2.txt -j test/sampletest2.json -t 1 -g 1 -f 43 -w 25 -v 0
 ./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211005_NapkenLk_duskdawn/20211005T124200-0400_SR.wav' -o test/sampletest2.txt -j test/sampletest2.json -t 1 -g 1 -f 43 -w 25 -v 1
make windDet
./windDet.exe -i test/windsample.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
 ./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211005_NapkenLk_duskdawn/20211005T124200-0400_SR.wav' -o test/sampletest2.txt -j test/sampletest2.json -t 1 -g 1 -f 86 -w 25 -v 1
 ./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211005_NapkenLk_duskdawn/20211005T124200-0400_SR.wav' -o test/sampletest2.txt -j test/sampletest2.json -t 1 -g 1 -f 86 -w 25 -v 1
make windDet
 ./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211005_NapkenLk_duskdawn/20211005T124200-0400_SR.wav' -o test/sampletest2.txt -j test/sampletest2.json -t 1 -g 1 -f 86 -w 25 -v 1
 ./windDet.exe -i /cygdrive/d/\!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20210826_NapkenLk_duskdawn/20210826T103500-0400_SR.wav -o test/sampletest3.txt -j test/sampletest3.json -t 1 -g 1 -f 86 -w 25 -v 1
./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211004_NapkenLk_duskdawn/20211004T203500-0400_SS.wav' -o test/Windless.txt -j test/Windless.json -v 1
./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211004_NapkenLk_duskdawn/20211004T203500-0400_SS.wav'  -j test/Windless2.json -f 500 -v 1
./windDet.exe -i '/cygdrive/d/!_TEMP_AUDIO_SAMPLES/ARU_RecordingSample_P7-04/20211004_NapkenLk_duskdawn/20211004T203500-0400_SS.wav' -o test/Windless2.txt -j test/Windless2.json -f 500 -v 1
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
which bash
./RunWindDet.sh 
./RunWindDet.sh 
d1='/cgddrive/d/TMP_R_Clones'
d2='WindNoiseDetection'
$d1$d2
cd $d1$d2
cd $d1/$d2
cd $d1\/$d2
d1='/cygddrive/d/TMP_R_Clones/'
cd $d1\/$d2
cd $d1$d2
d1='/cygdrive/d/TMP_R_Clones/'
cd $d1$d2
touch .gitignore
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
make windDet
./windDet.exe -i test/windsample.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
./windDet.exe -i test/iphone.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
make windDet
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
make windDet
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
make windDet
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 1
make windDet
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 0
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 0
make windDet
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 0
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
./RunWindDet.sh 
cd /cygdrive/d/TMP_R_Clones/WindNoiseDetection
make
make
make
make
make
./windDet.exe -i test/iphone1.wav -o test/iphone2.txt -j test/iphone2.json -t 1 -g 1 -f 43 -w 25 -v 0
make main_multi.cpp

src/processInWindows.cpp
/*Copyright (c) <2014> <Paul Kendrick>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
 */


#include "processInWindows.hpp"

struct wavfile {
    char id[4]; // should always contain "RIFF"
    int32_t totallength; // total file length minus 8
    char type[4]; // should always contain "WAVE "
    char typeId[4]; // should always contain "fmt "
    int32_t fmtchunksize; // format chunk size
    int16_t compression; // 1 pcm 3 ieee
    int16_t nochan; //
    int32_t fs; // fs
    int32_t bytes_per_sec; // byes / sec
    int16_t blockalign; //
    int16_t bits_per_sample; //
} __attribute__((__packed__));

static struct wavfile header;

int loadWav(const char *filename, const char * outFilename,const char *jsonFilename, char *treeDir, float gain, int frameAve,float thresh,int verbose,char * treeFileLoc) {
    char str1[100], str2[100];
    FILE * pFile;
    FILE *pFileJSON;


    sprintf(str1, "%s/%s/levelClass",treeFileLoc, treeDir);
    sprintf(str2, "%s/%s/snrClass",treeFileLoc, treeDir);
    DTree LevelTree, SNRTree;
            
    if (verbose ==1){printf("\nLoading Decision Trees");}
    LevelTree.readTextFilesTrees(str1);
    SNRTree.readTextFilesTrees(str2);
   if (verbose ==1){ printf("\nDone");}

    //http://www.sonicspot.com/guide/wavefiles.html
    //http://yannesposito.com/Scratch/en/blog/2010-10-14-Fun-with-wav/char 
    //char  filename[]="test16S.wav";
    //char  filename[]="iphone1.wav";

    FILE * wav;

    wav = fopen(filename, "r");
     if(wav==NULL)
     {
    if (verbose ==1){   printf("Can't read input file \n");}
        return 1;exit(1);
    }
       
    //check openable
    // int test = sizeof (header);
    if (fread(&header, sizeof (header), 1, wav) < 1) {
        fprintf(stderr, "Can't read input file header %s\n", filename);
       if (verbose ==1){  printf("Can't read input file header %s\n", filename);}

        return 1;exit(1);
    } 
    if (verbose ==1){
    printf("\nWavefile Header\n");
    printf("chunk ID %.*s\n", 4, header.id);
    printf("Total length %i\n", header.totallength);
    printf("should say WAVE %.*s\n", 4, header.type);
    printf("should say fmt : %.*s\n", 4, header.typeId);
    printf("format chunk size : %i\n", header.fmtchunksize);
    printf("Compression code %i\n", header.compression);
    printf("No. Channels %i\n", header.nochan);
    printf("Fs  %i \n", header.fs);
    printf("bytes_per_sec  %i \n", header.bytes_per_sec);
    printf("blockalign  %i\n", header.blockalign);
    printf("bits_per_sample  %i\n", header.bits_per_sample);
    }
    if (header.fs != 44100.0) {
        fprintf(stderr, "Input must be sampled at 44100 Hz \n");
        exit(0);
    }
    // Skip though rest of header to find data 
    char datahead[4];
    int32_t datasize;
    int skip = 0;
    while (skip == 0) {
        fread(&datahead, sizeof (datahead), 1, wav);
        //  Skips through unitll it finds 'data'
        skip = memcmp(datahead, "data", 4) == 0;
        if (skip == 0) {
            fseek(wav, -3, SEEK_CUR);
            //printf("should say data %.*s\n", 4, datahead);
        }
    }

    // read data header 
    // printf("should say data %.*s\n", 4, datahead);
    fread(&datasize, sizeof (int32_t), 1, wav);

    // printf("datasize %i\n", datasize);
    // printf("header.nochan %i\n", header.nochan);
    // printf("header.bits_per_sample %i\n", header.bits_per_sample);
    int32_t NoBlocks = datasize / (header.bits_per_sample / 8 * header.nochan);
    // printf("Number blocks  %i\n", NoBlocks);
    // allocate memory for time domain items
    float *window;
    float *windowOver;
    float *windowPrev;
    window = (float*) malloc(sizeof (float)*WIN_N);
    windowOver = (float*) malloc(sizeof (float)*WIN_N);
    windowPrev = (float*) malloc(sizeof (float)*WIN_N);

    float * mfcc = (float*) malloc((16) * sizeof (float));
    float * mfccO = (float*) malloc((16) * sizeof (float));
    float * mfccP = (float*) malloc((16) * sizeof (float));

    float *feats;
    feats = (float*) malloc(sizeof (float)*12 * 3);

    int wN = 0;
    int wNo = WIN_N / 2;

    sprintf(str2, "%s.b", outFilename);    remove(str2);
    pFile = fopen(str2, "w");
    


    //if the wav is 16 bit PCM
    int counter;
    int firstWin=0;
    int classLevel;
    int classSNR;
    int i;
    int frame = 0;
    int sumclassSNR = 0;
    int sumclassLevel = 0;
    float avermsAW = 0;
    initVarsFeats(WIN_N, header.fs);
    for (i = 0; i < (WIN_N); i++)
        window[i] = 0;

    if (1 == 1)//(header.bits_per_sample==16 && header.compression==1)
    {



        int i, c, n;
        for (i = 0; i < (NoBlocks); i++) {
            //channelsum[i]=0;
            for (c = 0; c < (header.nochan); c++) {
                float tmp2;
                float normV;
                if (header.bits_per_sample == 16 && header.compression == 1) {
                    int16_t tmp;
                    fread(&tmp, sizeof (int16_t), 1, wav);
                    tmp2 = (float) tmp;
                    normV = (65535.0 / 2.0);
                } else if (header.bits_per_sample == 32 && header.compression == 1) {
                    int32_t tmp = 0;
                    fread(&tmp, sizeof (int32_t), 1, wav);
                    tmp2 = (float) tmp;
                    normV = (2147483648.0);
                } else if (header.bits_per_sample == 32 && header.compression == 3) {
                    float tmp = 0;
                    fread(&tmp, sizeof (float), 1, wav);
                    tmp2 = tmp;
                    normV = 1; //(16777216.0/2.0);
                } else {
                    if (verbose ==1){
                   printf("Supported formats are 16 and 32 bit signed integer, and IEEE float little-endian - any number of channels");
                    };
                   exit(1); break;
                }
            if (feof(wav))
            {
               
                if (verbose ==1){
                    printf("End of file was reached unexpectedly.\n");
                }
                      break;//return 1;exit(1);

            }
            if (ferror(wav))
            {
                if (verbose ==1){
                printf("An error reading the wav occurred.\n");}
                return 1;exit(1);
            }
                float tmp3 = (tmp2 / (float) header.nochan) / normV * 158489.0 * gain; //*  158489=  10^(104/20)
                window[wN] = window[wN] + tmp3;
                // printf("%f \n",window[wN]);
                windowOver[wNo] = window[wN];

                //fprintf(pFile,"%f \n",channelsum[i]);
            } //65,535
            //close(pFile);

            //every WIN_N samples do this
            wN++;

            if (wN == WIN_N) {
                // printf("1 \n");

                for (n = 0; n < WIN_N / 2; n++)
                    windowOver[n] = windowPrev[n];
                for (n = (WIN_N) / 2; n < WIN_N; n++)
                    windowOver[n] = window[n];
                computeMFCCs(window, WIN_N, header.fs, mfcc);
                // for ( i=0;i<12;i++)         printf("%f \n",mfcc[i]);
                computeMFCCs(windowOver, WIN_N, header.fs, mfccO);
                computeMFCCs(windowPrev, WIN_N, header.fs, mfccP);


                for (n = 0; n < 36; n++) {
                    if (n < 12) {
                        feats[n] = mfccP[n];
                    } else if (n < 24) {
                        feats[n] = mfccO[n - 12];
                    } else {
                        feats[n] = mfcc[n - 24];
                    }
                    //  printf("%f \n",feats[n]);
                }

                if (frame > 3) {

                    classLevel = LevelTree.decisionTreeFun(feats);
                    classSNR = SNRTree.decisionTreeFun(feats);
                    sumclassLevel = sumclassLevel + classLevel;
                    sumclassSNR = sumclassSNR + classSNR;
                    avermsAW = avermsAW + pow(10, mfcc[0] / 10.0);
                    counter++;
                }
                int n;
                if ((frame % frameAve) == 0 && frame > 1) {

                    float avedBA = 10.0 * log10(avermsAW / (float) counter);
                    if (isinf(avedBA)==1)
                        avedBA=0;
                    float aveLevel = sumclassLevel / (float) counter;
                    float aveSNR = sumclassSNR / (float) counter;
                    float time = (float) WIN_N / header.fs * (float) frame - (float) WIN_N / header.fs / 2.0;
                    float tmp = 1;
                    float Combined = 0;
                    if (aveLevel > 1) {
                        tmp = 1.0 / 5.0;
                        Combined = (5 - aveSNR) * tmp;
                    }
                    if (verbose ==1){
                    printf("Time %0.2f dBA %0.0f AveWind %0.2f AveSNR %0.2f Combined %0.2f\n", time, avedBA, aveLevel, aveSNR, Combined);
                    }
                    fprintf(pFile, " %0.2f %0.0f %0.2f %0.2f %0.2f\n", time, avedBA, aveLevel, aveSNR, Combined);

                    sumclassLevel = 0;
                    sumclassSNR = 0;
                    counter = 0;
                    avermsAW = 0;
                    // printf("%f %i %i %f\n",mfcc[0],classLevel,classSNR,(float)(5-classSNR)/5.0*3.0*(float)classLevel);
                }

                //reset counters /vectors
                for (n = 0; n < 12; n++)
                    mfccP[n] = mfcc[n];
                for (n = 0; n < WIN_N; n++)
                    windowPrev[n] = window[n];
                wN = 0;
                frame++;
                for (n = 0; n < (WIN_N); n++)
                    window[n] = 0;                                     

            }



            // printf("%f \n",channelsum[i]);
            //fprintf(pFile,"%f \n",channelsum[i]);

        }
    }
    fclose(pFile);


    // Global stats
    float aveAveLevel = 0;
    float maxAveLevel = 0;
    float minAveSNR = 5;
    float aveAveSNR = 0;
    //float thresh = 0.2;
    float avecomb = 0;
    float maxcomb = 0;
    float count_wF5 = 0;
    float count_wF4 = 0;
    float count_wF3 = 0;
    float count_wF2 = 0;
    float count_wF1 = 0;
    float count_wF0 = 0;
    counter = 0;
    //  pFile2 = fopen ( 'tmp.txt' , "w");
    pFile = fopen(str2, "r");
    char mystring[1000];
    while (fgets(mystring, sizeof (mystring), pFile) != NULL) {

        float dBA, t, level, snr, comb;
        sscanf(mystring, "%f %f %f %f %f", &t, &dBA,
                &level, &snr, &comb);

        if (level > maxAveLevel)
            maxAveLevel = level;

        if (snr < minAveSNR)
            minAveSNR = snr;

        if (comb > maxcomb)
            maxcomb = comb;

        avecomb = avecomb + comb;
        aveAveSNR = aveAveSNR + snr;
        aveAveLevel = aveAveLevel + level;

        if (level < 1)
            count_wF0++;


        if (snr >= 0 && snr < 1 && level >= 1)
            count_wF5++;
        if (snr >= 1 && snr < 2 && level >= 1)
            count_wF4++;
        if (snr >= 2 && snr < 3 && level >= 1)
            count_wF3++;
        if (snr >= 3 && snr < 4 && level >= 1)
            count_wF2++;
        if (snr >= 4 && snr < 5 && level >= 1)
            count_wF1++;



        //printf("%f %f %f %f %f \n", t, dBA, 
        //       level,snr, comb);
        counter++;
    }               

    fclose(pFile);
    count_wF0 = count_wF0 / counter * 100.0;
    count_wF1 = count_wF1 / counter * 100.0;
    count_wF2 = count_wF2 / counter * 100.0;
    count_wF3 = count_wF3 / counter * 100.0;
    count_wF4 = count_wF4 / counter * 100.0;
    count_wF5 = count_wF5 / counter * 100.0;

    avecomb = avecomb / counter;
    aveAveSNR = aveAveSNR / counter;
    aveAveLevel = aveAveLevel / counter;
    
      remove(outFilename);
      remove(jsonFilename);

pFileJSON = fopen(jsonFilename, "w");    
//

    FILE * pFile2;
    pFile2 = fopen(outFilename, "w");
    pFile = fopen(str2, "r");
    fprintf(pFile2, "Microphone Wind Noise Detection - University of Salford - the Good Recording Project http://www.goodrecording.net \n\n");//, filename);
    fprintf(pFile2, "Microphone Wind noise Analysis for input file %s\n\n", filename);
    fprintf(pFile2, "Wind Noise Statistics, %% number of frames with wind noise detected at the following Signal to noise Ratios (high values = good quality low values = bad quality). also presented is the range in Quality Degradation that each SNR range represents.\n\n");
    fprintf(pFile2, "SNR range;\t>20dB,\t20 to 10dB,\t10 to 0dB,\t0 to -10dB,\t-10 to -20dB,\t<-20dB\n");//, count_wF0, count_wF1, count_wF2, count_wF3, count_wF4, count_wF5);
    fprintf(pFile2, "Qual. Degradation;\t<24%%,\t24 to 40%%,\t40 to 54%%,\t54 to 69%%,\t69 to 95%%,\t>95%%\n");
    fprintf(pFile2, "%% of time in each SNR range;\t%0.1f,\t%0.1f,\t%0.1f,\t%0.1f,\t%0.1f,\t%0.1f\n", count_wF0, count_wF1, count_wF2, count_wF3, count_wF4, count_wF5);
    fprintf(pFile2, "\n");
    fprintf(pFile2, "Wind noise time history\n\n");
    fprintf(pFile2, "T(s)\t\tQuality Degradation(%%)\t dBA \n");
    
    //JSON global stats
   //  fprintf(pFileJSON, "[\n\t{\n\t\"test\": \"Global Stats\",\n\t\"results\": [" );                    
    fprintf(pFileJSON, "{\n \t\"Global Stats\":\n \t [ \n" );                    
    fprintf(pFileJSON, "\t\t%0.1f,\t%0.1f,\t%0.1f,\t%0.1f,\t%0.1f,\t%0.1f\n\t\t],\n", count_wF0, count_wF1, count_wF2, count_wF3, count_wF4, count_wF5);
    fprintf(pFileJSON, "\t \"FileName\": \"%s\",\n", filename);
    fprintf(pFileJSON, "\t \"Time History\":\n\t [\n" );  
    float tLast=0;
    while (fgets(mystring, sizeof (mystring), pFile) != NULL) {

        float dBA, t, level, snr, comb;
        sscanf(mystring, "%f %f %f %f %f;", &t, &dBA,
                &level, &snr, &comb);
        float x=snr;
        float qual= 0.0093616465*pow(x,9) + -0.2408574076*pow(x,8) + 2.6409491876*pow(x,7) + -16.0598328947*pow(x,6) + 59.1080787296*pow(x,5) + -135.0673590478*pow(x,4) + 189.6816801547*pow(x,3) + -156.8624014598*pow(x,2) + 83.7429623557*pow(x,1) + 4.5226369050;
        if (comb==0)
            qual=100;

        fprintf(pFile2, "%4.1f\t\t%0.0f\t\t\t\t\t\t%4.0f\n", t, 100-qual, dBA);
                    if (firstWin ==1)
                      fprintf(pFileJSON, ",\n");                   
                    fprintf(pFileJSON, "\t\t{\"Ts\": %0.2f,\"Te\": %0.2f, \"dBA\": %0.0f, \"QDeg\": %0.2f}", tLast,t, dBA, 100-qual);                    
                    firstWin=1;
                    tLast=t;
        counter++;
    }

    ///////////////////////////// Find contiguous regions without wind noise
    fclose(pFile);
    pFile = fopen(str2, "r");
    fprintf(pFileJSON, "\n\t],\n");
    fprintf(pFileJSON, "\t\"Wind free regions\":\n \t [\n" );                    

    firstWin=0;
    int clean = 1; // 
    int initialised = 0;
    int first = 0;
    float win = 0;
    float start = 0;
    fprintf(pFile2, "\nWind free regions from - to (s) using a Threshold of %2.0f\n\n",thresh);
   float dBA, t, level, snr, comb;
    while (fgets(mystring, sizeof (mystring), pFile) != NULL) {

        
        
        sscanf(mystring, "%f %f %f %f %f", &t, &dBA,
                &level, &snr, &comb);
               float x=snr;
        float qual= 0.0093616465*pow(x,9) + -0.2408574076*pow(x,8) + 2.6409491876*pow(x,7) + -16.0598328947*pow(x,6) + 59.1080787296*pow(x,5) + -135.0673590478*pow(x,4) + 189.6816801547*pow(x,3) + -156.8624014598*pow(x,2) + 83.7429623557*pow(x,1) + 4.5226369050;
        if (comb==0)
            qual=100;
    comb=100-qual;
        
        if (comb >= thresh && clean == 1 && initialised == 1 && first == 1) // previous frame was clean now its windy
        {
            fprintf(pFile2, "%0.2f\t%0.2f\n", start, t - win * 2);
                   if (firstWin ==1)
                      fprintf(pFileJSON, ",\n");                   
                    fprintf(pFileJSON, "\t\t{\"s\": %0.2f, \"e\": %0.2f}", start, t - win * 2 );                    
                    firstWin=1;
            
            clean = 0;

        } else if (comb < thresh && clean == 0 && initialised == 1) // previous frame was windy now its clean
        {
            start = t - win * 2;
            clean = 1;
            first = 1;
        }

        if (initialised == 0) {
            win = t / 2;

            if (comb >= thresh) {
                clean = 0;

            } else {
                clean = 1;
                first = 1;
                start = t - win * 2;
            }
            initialised = 1;
        }
        counter++;
    }
    ///////////////////////////
 if (comb < thresh && clean == 1 ) // last frame if clean
        {
            fprintf(pFile2, "%0.2f\t%0.2f\n", start, t+1);
                   if (firstWin ==1)
                      fprintf(pFileJSON, ",\n");                   
                    fprintf(pFileJSON, "\t\t{\"s\": %0.2f, \"e\": %0.2f}", start, t  );                    
                    firstWin=1;
            
            clean = 0;

 }
    fclose(pFile);
    fclose(pFile2);
     fprintf(pFileJSON, "\n\t]\n}\n");
     fclose(pFileJSON);

    remove(str2);

    return 1;
}

